File: 799-champagne-tower/799-champagne-tower.py
class Solution:
    def champagneTower(self, poured: int, query_row: int, query_glass: int) -> float:
        rest = lambda glass: max(0, (glass-1)/2)
        row = [0, poured, 0]
        for i in range(query_row):
            row = [0] + [rest(row[j]) + rest(row[j+1]) for j in range(i+2)] + [0]
        return min(1, row[query_glass+1])

File: 799-champagne-tower/799-champagne-tower.java
class Solution {
    public double champagneTower(int poured, int query_row, int query_glass) {
        double[][] pascal = new double[101][101];
        pascal[0][0] = poured;
        
        for(int i=0;i<=query_row;i++){
            for(int j=0;j<=query_glass;j++){
                if(pascal[i][j] > 1.0){
                    pascal[i+1][j] += (pascal[i][j] - 1.0) / 2.0;
                    pascal[i+1][j+1] += (pascal[i][j] - 1.0) / 2.0;
                    pascal[i][j] = 1;
                }
            }
        }
        
        return pascal[query_row][query_glass];
    }
}